app/models/amazon_product_search.rb

```diff
@@ -33,7 +28,17 @@ def upcs_list_to_string
     upcs.join(",")
   end
 
+  private
+
+  def all_product_data
+    @_all_product_data ||= service.get_info(upcs_list_to_string)
+  end
+
```

Comment:
I moved `all_product_data` to a private method (I guess it doesn't really need to be private, but whatever). I'm memoizing it here with the `||=` operator. That way, when we call `all_product_data` in the `unique_product_data` method, we aren't making repeat calls to Amazon.